[WIP] Fix Floats comparisons

asmeurer opened the pull request:
This uses evalf, which is the correct logic for high precision Floats.

Still needs to add tests, fix other comparison methods in numbers.py, and fix
some test failures.

This is issue #12534 (dup #12327).

asmeurer committed:
Compare Floats using is_positive

This uses evalf, which is the correct logic for high precision Floats.

Still needs to add tests, fix other comparison methods in numbers.py, and fix
some test failures.

This is issue #12534.

jksuom commented:
It seems that some additional checking is needed in Float - nonreal comparisons.
```
>>> S(0) < I
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/core/numbers.py", line 1960, in __lt__
    return Rational.__lt__(self, other)
  File "sympy/core/numbers.py", line 1618, in __lt__
    return Expr.__lt__(expr, other)
  File "sympy/core/expr.py", line 321, in __lt__
    raise TypeError("Invalid comparison of complex %s" % me)
TypeError: Invalid comparison of complex I
>>> S(0.0) < I
False
```

asmeurer commented:
Ah, of course. `I.is_positive` gives False.

asmeurer committed:
Float comparisons only test against Rationals and other Floats

The more complicated logic lives in the Expr comparison methods. Them
themselves recursively call these when they reduce the problem to a Float
comparison.

asmeurer commented:
OK, this still needs tests, and I need to fix the Rational methods, but I think the current fix is better than what I had. The evalf logic is kept in `Expr.__lt__`, etc. only (where it is already implemented). `Float.__lt__`, etc. only implements comparison against other rational and floating point numbers. 

The more general `evalf(2)` logic is still implemented redundantly in at least six places (`Expr.__gt__`, `Expr__ge__`, `Expr.__lt__`, `Expr.__le__`, `Expr._eval_is_positive`, and `Expr.__eval_is_negative`). I don't really feel like sorting out the differences there and deduplicating the code. As far as I can tell, that code (in `Expr`) as it currently is is correct, so I'm not too motivated to change it.

asmeurer committed:
Rational comparison methods only compare against other rationals

We have to manually dispatch for Float and NumberSymbol because NotImplemented
breaks some code in relational.py that explicitly avoids that logic to prevent
symbolic inequalities from being flipped.

asmeurer commented:
And also an implementation for Rational. I ended up removing some code from 2ccd074e918 because it didn't seem necessary. 

Some tests are failing that were wrong (saying that `a = S("6506833320952669167898688709329/5070602400912917605986812821504")` and `a.n(10)` should be equal). So I'm actually a little concerned that this may indeed spill over and require fixing https://github.com/sympy/sympy/issues/11707. But maybe I can just ignore == and != for now and punt until we have a fix for that issue, even if it means `==` and `<` are inconsistent.

jksuom commented:
It also appears that Float - Rational comparisons should be improved. Tests like
```
        if isinstance(other, Rational):
            return Rational(self).__lt__(other)
```
fail because of the conversions
```
>>> Rational(Float('inf'))
0
>>> Rational(Float('-inf'))
0
```

smichr commented:
Leave a comment

> The more general evalf(2) logic is still implemented redundantly in at least six places

I'll see if I can get that cleaned up then send a PR to this PR if you haven't already committed this by the time I do so.

asmeurer commented:
As I said, I have no plans to do any cleanups for the Expr code, unless I find bugs in it, so go ahead. Hopefully PRing this branch isn't too difficult, given the test failures that currently exist on it. If so, you can make a separate PR to master and I can worry about merge conflicts if there are any (there shouldn't be).

asmeurer commented:
By the way, @smichr, I've noticed that some of your GitHub comments start with "leave a comment" (on other issues too).

asmeurer commented:
The tests hang/crash in the `test_evalf_complex_powers` test in `test_evalf.py`. The issue comes from trying to convert the float from `NS('(E+pi*I)**100000000000000000')` into a Rational (`NS(...)` = `sympify(...).evalf()`). The `2**expt` line tries to create a huge number. So I guess we have to be a little smarter about comparisons with rationals.

jksuom commented:
Maybe we should convert the Rational `other` into a Float for comparison.

asmeurer commented:
More than likely the same `(a - b).evalf(2) > 0` logic will work with rationals as well. I just need to restructure the code, since Expr calls literally "`(a - b).evalf(2) > 0`", which is itself a `Float > Rational`. It should be enough to only special-case S.Zero, instead of all rationals (testing if a float is positive or negative is trivial).

jksuom commented:
`test_evalf_complex_powers` in test_evalf fails at
`NS('(E+pi*I)**100000000000000000')` where
```
def NS(e, n=15, **options):
    return sstr(sympify(e).evalf(n, **options), full_prec=True)
```
There is no problem with computing
```
>>> e = '(E+pi*I)**100000000000000000'
>>> sympify(e).evalf(15)
-3.58896782867793e+61850354284995199 + 4.58581754997159e+61850354284995199*I
```
but the printer will call
```
  File "sympy/printing/precedence.py", line 73, in precedence_Float
    if item < 0:
  File "sympy/core/numbers.py", line 1153, in __lt__
    return Rational(self).__lt__(other)
```
with `self = item = -3.58896782867793e+61850354284995199`, and no available memory will suffice for the numerator of the requested rational fraction.

asmeurer committed:
Only compare floats directly against zero

Previously, we compared against all rationals by converting to Rational first.
But computing the exact rational representation is impossible for floats with
extremely large exponents.

Instead, pass all nontrivial comparisons to Expr.__lt__ (etc.), which uses
(a - b).evalf(2) > 0, and only implement the comparison against 0 so that that
recursive call works.

asmeurer committed:
Remove evalf() call in NumberSymbol.__le__

This is invalid because it only evalfs one side of the inequality (like
pi.evalf() <= 0), which is not enough digits if the right hand side is higher
than the default precision (like pi <= pi.evalf(100)).

Instead, fallback to Expr.__lt__, which computes (self - other).evalf(2).

asmeurer commented:
OK, much better fix now, I hope. I want to see if the tests pass, and if they do, I will add some tests for the fix itself, and then this can be reviewed.

asmeurer commented:
Oh I still need more cleanups of the NumberSymbol inequalities.

asmeurer commented:
It seems `evalf(2)` isn't as smart as I would have hoped. 

```py
>>> (pi - Rational(pi.evalf(110))).evalf(2)
0.e-110
>>> (pi - Rational(pi.evalf(110))).evalf(2)._mpf_
(0, mpz(1), -365, 1)
>>> (pi - Rational(pi.evalf(110))).evalf(2).is_comparable
False
>>> (pi - Rational(pi.evalf(110))).evalf(2)._prec
1
```

which causes `pi > Rational(pi.evalf(110))` to say unevaluated in this branch. If I bump it up to `evalf(3)` it works.  

```py
>>> (pi - Rational(pi.evalf(110))).evalf(3)
1.7e-111
```

But if you increase the `110` part you need higher and higher precision:

```
>>> (pi - Rational(pi.evalf(200))).evalf(64)
0.e-201
>>> (pi - Rational(pi.evalf(200))).evalf(65)
1.42e-201
```

We also get this behavior:

```
>>> (pi.evalf() - Rational(pi.evalf(200))).evalf(2)
0
```

Which is clearly not correct

asmeurer commented:
The last one is because `Float.__sub__` evalfs the other expression with the `self._prec` if it is a Number, that is, `pi.evalf() - Rational(pi.evalf(200))` gives 0 directly.

asmeurer commented:
OK, so `_as_mpf_op` tries to find a common precision when adding two Number objects, using `max`: https://github.com/sympy/sympy/blob/39bbad5a86213c6191b26cbef13310770c544d1a/sympy/core/numbers.py#L435. So the logic is right for Float - Float:

```
>>> (pi.evalf() - pi.evalf(200))
-0.00000000000000012246467991473531772260659322750010582097494459230781640628620899862803482534211706798214808651328230664709384460955058223172535940812848111745028410270193852110555964462294895493038196300946148827247
```

But the problem is that Rational._prec is set as -1:

```
>>> Rational(1, 2)._prec
-1
```

In actuality, you would want Rational to have infinite precision relative to a Float. 

It might be possible to define `Rational._as_mpf_op` to give a precision for the Rational relative to a Float? Would that depend on the operation (add, sub, mul, div, mod)? I'll need to dig into the evalf logic to see if this is already implement (someone let me know if you already know).

jksuom commented:
The result of `evalf(2)` can also be improved by raising the maximum working precision:
```
>>> (pi - Rational(pi.evalf(110))).evalf(2, maxn=110)
1.7e-111
>>> (pi - Rational(pi.evalf(200))).evalf(2, maxn=200)
1.4e-201
```

asmeurer commented:
I'm not too worried about certain inequalities like `pi < pi.evalf(200)` not evaluating without a higher maxn. It would be nice if there were some obvious way users could make them evaluate, but that's a separate issue. The code (as far as I can tell) should be correct, in that it won't evaluate to a wrong answer when there isn't enough precision. 

But I do think that every `Number < Number` should evaluate, in particular, `Float < Rational`. If we have `m/2**e < p/q`, this is equivalent to `q*m < p*2**e`, and we need to evaluate this, critically, without evaluating `2**e`, as `e` could be large. 

I think one way to do it would be to find the power of 2, `2**n`, such that `2**n <= q*m/p < 2**(n+1)`, and then compare `n < e`. I don't know if this is the most efficient way. Are there already fast routines in SymPy/mpmath to bound a number by consecutive powers of 2?

jksuom commented:
mpmath's comparisons are fairly efficient. I think we should use them also in Float -- Rational comparisons.  To compare a Float `a` with a Rational `p/q` we only need to compute `q*a`, which is more efficient than converting `p/q` to a Float, and then compare `q*a` with `Float(p)`.

asmeurer commented:
Thanks, that's much simpler.

asmeurer commented:
Going to have to postpone this for the release. If anyone wants to pick up the work, feel free.

smichr commented:
I can't see enough context to tell why you deleted the code handling NumberSymbol in one place but not another. The code for NumberSymbol comparisions is being corrected with #13429 and the approach with `approximation_interval` is sound (it just tries to handle the simple cases quickly) so I think you can leave that in place in all cases -- but again, maybe I am missing something.

asmeurer commented:
I'll try to take a look. Probably won't get to it until at least next week though. Please remind me if I don't after next week.

asmeurer commented:
The code from approximation_interval does look correct now that I've given it a second look. I guess I didn't realize that it's only used for checking *outside* the interval. The bug is the line right below that in `NumberSymbol.__lt__`, where it calls `_sympify(self.evalf() < other)` when the approximation_interval check fails. The `self.evalf()` call can make `self` bigger than `other` when it shouldn't be if the 15 digit evalf rounds that way. 

Either way, I would do some tests once we have a proper fix if approximation_interval is even needed. It seems like a premature optimization.

asmeurer commented:
As for deleting code in one place but not another, I think the main reason this is still WIP is that I didn't get around to fixing all the code yet (from what I remember). That, and I still need to implement the proper comparison for rationals as suggested by Kalevi above.

smichr commented:
I implemented the Float-Rational method suggested and removed use of the approximation interval (which, at worst, would require 2 additional comparisions).

asmeurer commented:
Another thought. The heuristics (IIRC), for something like `(pi - Rational).evalf(2)` double the precision up to some point. But given the Rational we can compute an exact bound on the number of digits potentially needed. Perhaps this is already implemented? My worry is very large rationals not evaluating.  

Also, in the case of a very large rational, the approximation interval actually would be a good optimization, because Rational < Rational is much faster than pi.evalf(large_N).  

Ditto for very precise Float instead of very large rational (although those are much rarer, as you have to manually evalf() to a large precision and most people don't do that).

czgdp1807 commented:
@asmeurer Probably this looks like an important PR as it is delayed from previous few versions. I am giving it higher priority,

asmeurer commented:
Some of these issues might have already been fixed by other pull requests. I'm not sure what the status of this is at this point in time. CC @smichr

czgdp1807 commented:
Okay, I will test all the code given in the associated issue to check if they are fixed. Thanks for the update.

czgdp1807 commented:
Some updates on the latest status. I have also added in the comments, what I would have expected as a user.
```python
Python 3.6.8 (default, Aug 20 2019, 17:12:48) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> Rational(pi.evalf(100)) > pi # correct result
False
>>> Rational(pi.evalf(101)) > pi # isn't it incorrect result?
True
>>> Rational(pi.evalf(100)) < pi # correct
True
>>> Rational(pi.evalf(101)) < pi # incorrect?
False
>>> pi.evalf(100) > pi # correct
False
>>> pi.evalf(100) < pi # correct
True
>>> pi.evalf(100) > pi.evalf(101) # correct
False
>>> pi.evalf(100) < pi.evalf(101) # correct
True
>>> (pi.evalf(100) - pi).is_positive # correct
False
>>> (pi - pi.evalf(100)).is_positive # correct
True
>>> (pi.evalf(100) - pi.evalf(101)).is_positive # correct
False
>>> (pi.evalf(101) - pi.evalf(100)).is_positive # correct
True
>>> S('0.3') == Rational(3, 10) # incorrect?
False
>>> hash(S('0.3')) == hash(Rational(3, 10)) # incorrect?
False
>>> [S('0.3')] == [Rational(3, 10)] # incorrect?
False
>>> {S('0.3')} == {Rational(3, 10)} # incorrect?
False
>>> S("0.3") in [Rational(3, 10)] # incorrect?
False
>>> S("0.3") in {Rational(3, 10)} # incorrect
False
```
ping @smichr @asmeurer

asmeurer commented:
S("0.3") produces Float("0.3"), which is not the same as 1/3:

```py
>>> Rational(Float("1.3"))
5854679515581645/4503599627370496
```

Regarding the evalfs, it depends on if the 100th or 101st digits of pi round up or down (using binary floating point rounding).

czgdp1807 commented:
> S("0.3") produces Float("0.3"), which is not the same as 1/3

Is it the expected behaviour?

> Regarding the evalfs, it depends on if the 100th or 101st digits of pi round up or down (using binary floating point rounding).

Are we okay with this, or some fix is needed for this?

asmeurer commented:
> Is it the expected behaviour?

Yes.

> Are we okay with this, or some fix is needed for this?

You need to check what the 100th and 101th digits are. I think I specified above somewhere what the correct result should be. A simple way to test is to check if `Rational(pi.evalf(100))` gives the same answer, since Rational(Float) gives the exact rational representation of the float, and we know rational comparisons aren't broken.

jksuom commented:
It seems that some (all?) of the issues in https://github.com/sympy/sympy/issues/12534 have been fixed, probably by https://github.com/sympy/sympy/pull/16770.

oscarbenjamin commented:
@asmeurer what is the status of this?

oscarbenjamin commented:
This is currently on the 1.5 milestone

asmeurer commented:
I think the important things have been fixed. It can be removed from milestone. 

We still need to go through this and the issues to see if any outstanding float issues still remain.
